### PR TITLE
chore: Make site-admin checks part of GraphQL API

### PR DESCRIPTION
**Pull request by varungandhi-src**
We can utilize the new directives support added in graphql-go
to mark which fields must be accessible only to site-admins
in the GraphQL API. Then we can use the middleware to check for
these, instead of having to add the checks to code.

## Test plan

TODO: Add new tests. What is the easiest way to test this?

### CODE REVIEW

File: cmd/frontend/auth/non_public.go [hunk 1]
```diff
@@ -20,39 +26,65 @@ import (
...
+		},
+	}
+}
+
+func GraphQLAuthMiddleware(schema *graphql.Schema) *Middleware {
+	return &Middleware{
+		API: func(next http.Handler) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				handler := relay.Handler{Schema: schema}
+				handler.ServeHTTP(w, r)
```
Comment (varungandhi-src) [inline cmd/frontend/auth/non_public.go:79]: Ah, this is not correct, this will start executing the request directly.

**Review by evict — APPROVED** [review]
Very nice idea! LGTM from security POV. Makes adding checks a lot easier in the future too. :+1:  Thanks!

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 2]
```diff
@@ -3,6 +3,20 @@ schema {
...
+GraphQLRole is the role for checking API access.
+
+As of July 26 2024, not all fields which require site admin access
+have been marked with @hasRole(role: SITE_ADMIN), but we want
+to eventually migrate all fields.
+"""
+enum GraphQLRole {
+    USER
+    SITE_ADMIN
+}
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/schema.graphql:16]: I love the idea of this, but the term `Role` here is too overloaded I think. 

Since the introduction of RBAC in Sourcegraph, `Role` is an entity of it's own, and those are dynamically defined by the admin. 

I think we should either tie this to RBAC (and not rely on roles which are runtime-defined and instead rely on RBAC permissions),
or reword this to `requiresSiteAdmin` or so for now, to be clearer. I'd love to push for more RBAC adoption 😬

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 3]
```diff
@@ -3,6 +3,20 @@ schema {
...
+
+directive @hasRole(role: GraphQLRole!) on FIELD_DEFINITION
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/schema.graphql:18]: The `hasRole` works well for things where the whole function body is governed for site-admin access, but it doesn't cover well if under a specific condition site-admin is required.

Example:
```
func (r *resolver) SecretField(ctx context.Context) (string, error) {
  if err := auth.CheckCurrentUserIsSiteAdmin(ctx); err != nil { return "", err } // This line can be removed here since GQL governs it
  return "a secret", nil
}

func (r *userResolver) SecretField(ctx context.Context) (string, error) {
  if r.user.SiteAdmin {
    // Only site-admins can see secret values of site-admin users:
    if err := auth.CheckCurrentUserIsSiteAdmin(ctx); err != nil { return "", err } // This cannot be expressed in the GQL schema definition.
  }
  return "a secret", nil
}
```

Do we think it could be confusing / inconsistent that _some_ site-admin constraints can be defined in the schema, but others cannot? 

I haven't made up my mind about this yet, just wanted to call it out and start a discussion. 

Another thing I could see here is that since schema and implementation aren't closely tied together, there might be some odd code-reuse cases now where we might leak data. (or we decide to keep site-admin checks ALSO in code, which would be duplication then, and potentially get out of sync over time).

For example, if a resolver does this today:

```
func (r *repoResolver) SecretField(ctx context.Context) (string, error) {
 // No enforcement here, GQL layer handles it with the annotation on `secretField` in the schema.
  return "sosecret", nil
}

func (r *commitResolver) PublicField(ctx context.Context) (string, error) { // this field has NO annotation
  return r.repoResolver.SecretField(ctx) // Here is no enforcement since we're 1 layer below the GQL field resolution here, nothing guards you from calling and returning the value from `SecretField` which will leak it then. In the current implementation, `SecretField` would call `auth.CurrentUserIsSiteAdmin` which is safe regardless of the codepath.
}
```

File: cmd/frontend/gqlauth/gqlauth.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+const GraphQLRoleKey contextKey = iota
+
+func (h *HasRoleDirective) Validate(ctx context.Context, _ interface{}) error {
+	if h.Role != SiteAdminGraphQLRole {
+		return nil
+	}
+	if role, ok := ctx.Value(GraphQLRoleKey).(GraphQLRole); ok && role == SiteAdminGraphQLRole {
+		return nil
+	}
+	return errors.New("Operation requires site-admin permissions")
```
Comment (eseliger) [inline cmd/frontend/gqlauth/gqlauth.go:35]: I think we should return `auth.ErrMustBeSiteAdmin` here for consistency, but if you think this is not enough context we could also return `errors.Wrap(auth.ErrMustBeSiteAdmin, "field %s failed to query")`or so.

**Review by eseliger — COMMENTED** [review]
I like the idea and the free documentation of this, I started a few discussions in comments.

**Comment by varungandhi-src** [comment]
@eseliger sorry I should've communicated the status of this earlier. The graphql-go upgrade is triggering some error, so I'm going to close this PR for now. I'll address your feedback should we decide to go down this route in the future.

More context: https://sourcegraph.slack.com/archives/CHXHX7XAS/p1721972057663129?thread_ts=1721916870.294159&cid=CHXHX7XAS
